lib/Log/Any/Adapter/DERIV.pm

```diff
@@ -656,6 +659,57 @@ sub clear_context {
     $adapter_context= undef;
 }
 
+=head2 mask_sensitive
+
+Mask sensitive data in the message and logs error in case of failure
+
+=over 4
+
+=item * C<$message> string - The message to be masked
+
+=back
+
+Returns string - The masked message
+
+=cut
+
+sub mask_sensitive {
+    my ($message) = @_;
+
+    # Define a lookup hash reference for all sensitive data regex patterns to be logged
+    my $sensitive_patterns = {
+        'PII' => [
+            qr/\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b/,  # Email
+        ],
+        'Sensitive' => [
+            qr/\b(?:token|key|oauth[ _-]?token)\s*[:=]\s*([^\s]+)/i, # token or api key
+        ],
+        'Token' => [
+            qr/(?:a1|r1|ct1)-[a-zA-Z0-9]{29}/,  # OAuth, Refresh, and CTrader token patterns
+            qr/[a-zA-Z0-9]{15}/,                  # API Token pattern
+        ],
+    };
+
+    try {
+        foreach my $category (keys %$sensitive_patterns) {
+            foreach my $pattern (@{$sensitive_patterns->{$category}}) {
+                $message =~ s/$pattern/'*' x length($&)/ige;  
+            }
+        }
+    } catch ($e) {
+        # Disable the custom warning handler temporarily to avoid potential recursion issues.
+        local $SIG{__WARN__} = undef;
+
+        # Extract the error message from the exception.
+        chomp(my $error_msg = $e);
+
+        # Log the error for further investigation and troubleshooting.
+        $log->warn("Error in mask_sensitive: $error_msg");
```

Comment:
@chylli-deriv can you please confirm this wont introduce recursive loop cycle? I have read about this and seem we already using but just wanted to double check

Reply:
I think so. But you can always try at least a manual test.

Reply:
this is done @chylli-deriv